python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -636,11 +636,69 @@ def get_root_unreconciled_ancestors(
 
         return root_unreconciled_ancestors
 
-    def is_reconciled(self, asset_partition: AssetKeyPartitionKey) -> bool:
-        """Returns a boolean representing if the given `asset_partition` is currently reconciled.
-        An asset (partition) is considered unreconciled if any of:
-        - It has never been materialized
-        - One of its parents has been updated more recently than it has
-        - One of its parents is unreconciled.
+    @cached_method
+    def get_ancestors_missing_or_with_unreconciled_children(
```

Comment:
re: our conversation yesterday, we'd want to replace this with something that returns the immediate parents that have been updated more recently. In terms of factoring things out, I think the body of the `get_missing_or_updated_parents` function could be shared between this and the recursive function. I was toying around with it and it turned out to be a bit tougher than I thought it'd be, but I believe this should basically do the trick, although it'd need to be cleaned up a bit:

```python
    def _get_updated_or_missing_parent_asset_partitions(
        self,
        asset_partition,
        parent_asset_partitions,
    ):
        parent_asset_partitions_by_key: Dict[AssetKey, List[AssetKeyPartitionKey]] = defaultdict(
            list
        )
        for parent in parent_asset_partitions:
            parent_asset_partitions_by_key[parent.asset_key].append(parent)

        time_or_dynamic_partitioned = isinstance(
            self.asset_graph.get_partitions_def(asset_partition.asset_key),
            (TimeWindowPartitionsDefinition, DynamicPartitionsDefinition),
        )
        updated_or_missing_parents = set()
        for parent_key, parent_asset_partitions in parent_asset_partitions_by_key.items():
            # ignore non-observable source parents
            if self.asset_graph.is_source(parent_key) and not self.asset_graph.is_observable(
                parent_key
            ):
                continue

            # find missing parents
            for parent in parent_asset_partitions:
                if not self.asset_partition_has_materialization_or_observation(parent):
                    updated_or_missing_parents.add(parent)

            # when mapping from time or dynamic downstream to unpartitioned upstream, only check
            # for existence of upstream materialization, do not worry about updates
            if time_or_dynamic_partitioned and not self.asset_graph.is_partitioned(parent_key):
                continue

            updated_or_missing_parents.update(self.get_asset_partitions_updated_after_cursor(
                asset_key=parent_key,
                asset_partitions=parent_asset_partitions,
                after_cursor=self.get_latest_materialization_or_observation_storage_id(
                    asset_partition
                ),
            ))
        return updated_or_missing_parents

    @cached_method
    def get_root_unreconciled_ancestors(
        self, *, asset_partition: AssetKeyPartitionKey
    ) -> AbstractSet[AssetKey]:
        """Return the set of root unreconciled ancestors of the given asset partition, i.e. the set
        of ancestors of this asset partition which are unreconciled for a reason other than that
        one of their ancestors is unreconciled.
        """
        # always treat source assets as reconciled
        if self.asset_graph.is_source(asset_partition.asset_key):
            return set()
        elif not self.asset_partition_has_materialization_or_observation(asset_partition):
            return {asset_partition.asset_key}

        parent_asset_partitions = self.asset_graph.get_parents_partitions(
            dynamic_partitions_store=self,
            current_time=self._evaluation_time,
            asset_key=asset_partition.asset_key,
            partition_key=asset_partition.partition_key,
        ).parent_partitions

        updated_or_missing_parent_asset_partitions = self._get_updated_or_missing_parent_asset_partitions(
            asset_partition,
            parent_asset_partitions
        )

        root_unreconciled_ancestors = { asset_partition.asset_key } if updated_or_missing_parent_asset_partitions else set()

        # recurse over parents
        for parent in set(parent_asset_partitions) - updated_or_missing_parent_asset_partitions:
            root_unreconciled_ancestors.update(
                self.get_root_unreconciled_ancestors(asset_partition=parent)
            )
        return root_unreconciled_ancestors

```

One small note is that I think technically the missing / updated asset partitions should be split out into two different sets, as we should only count an asset_partition as a "root unreconciled asset partition" if its parent has been updated more recently than it. If its parent is missing, we should instead count the missing parent as the "root unreconciled asset partition". Minor change in code, just was too lazy to do it here

Reply:
Might also want to refactor how we're passing in the parent asset partitions, but a big thing here is that

```python
self.asset_graph.get_parents_partitions(...)
```

Can be a fairly expensive call so ideally we'd only do that once per asset partition